File: app/src/main/java/com/soundcloud/android/profile/UserDetailsFragment.java
package com.soundcloud.android.profile;

import com.soundcloud.android.R;
import com.soundcloud.android.SoundCloudApplication;
import com.soundcloud.android.analytics.ScreenProvider;
import com.soundcloud.android.analytics.SearchQuerySourceInfo;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.navigation.Navigator;
import com.soundcloud.android.presentation.RefreshableScreen;
import com.soundcloud.android.utils.LeakCanaryWrapper;
import com.soundcloud.android.view.MultiSwipeRefreshLayout;
import com.soundcloud.lightcycle.LightCycle;
import com.soundcloud.lightcycle.LightCycleSupportFragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import javax.inject.Inject;

public class UserDetailsFragment extends LightCycleSupportFragment<UserDetailsFragment>
        implements RefreshableScreen {

    @Inject LeakCanaryWrapper leakCanaryWrapper;
    @Inject UserProfileOperations profileOperations;
    @Inject Navigator navigator;
    @Inject ScreenProvider screenProvider;
    @Inject @LightCycle UserDetailsPresenter userDetailsPresenter;

    public static UserDetailsFragment create(Urn userUrn,
                                             SearchQuerySourceInfo searchQuerySourceInfo) {
        final UserDetailsFragment userDetailsFragment = new UserDetailsFragment();
        Bundle args = new Bundle();
        args.putParcelable(ProfileArguments.USER_URN_KEY, userUrn);
        args.putParcelable(ProfileArguments.SEARCH_QUERY_SOURCE_INFO_KEY, searchQuerySourceInfo);
        userDetailsFragment.setArguments(args);
        return userDetailsFragment;
    }

    public UserDetailsFragment() {
        SoundCloudApplication.getObjectGraph().inject(this);
        setRetainInstance(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.user_info_view, container, false);
    }

    @Override public void onDestroyView() {
        super.onDestroyView();
        leakCanaryWrapper.watch(this);
    }

    @Override
    public MultiSwipeRefreshLayout getRefreshLayout() {
        return (MultiSwipeRefreshLayout) getView().findViewById(R.id.str_layout);
    }

    @Override
    public View[] getRefreshableViews() {
        final View view = getView();
        return new View[]{view.findViewById(android.R.id.empty),
                view.findViewById(R.id.user_details_holder)};
    }
}

File: app/src/main/java/com/soundcloud/android/profile/UserDetailsFragment.kt
package com.soundcloud.android.profile

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.soundcloud.android.R
import com.soundcloud.android.SoundCloudApplication
import com.soundcloud.android.analytics.SearchQuerySourceInfo
import com.soundcloud.android.main.Screen
import com.soundcloud.android.model.AsyncLoadingState
import com.soundcloud.android.model.Urn
import com.soundcloud.android.presentation.RecyclerItemAdapter
import com.soundcloud.android.utils.Urns
import com.soundcloud.android.utils.collection.AsyncLoaderState
import com.soundcloud.android.view.BaseFragment
import com.soundcloud.android.view.collection.CollectionRenderer
import com.soundcloud.android.view.collection.CollectionRendererState
import dagger.Lazy
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject

class UserDetailsFragment : BaseFragment<UserDetailsPresenter>(), UserDetailsView {

    override val presenterKey: String = "UserDetailsPresenterKey"
    override val linkClickListener: PublishSubject<String> = PublishSubject.create()
    override val followersClickListener: PublishSubject<UserFollowsItem> = PublishSubject.create()
    override val followingsClickListener: PublishSubject<UserFollowsItem> = PublishSubject.create()

    @Inject lateinit var presenterLazy: Lazy<UserDetailsPresenter>
    @Inject lateinit var adapterFactory: UserDetailAdapter.Factory

    private lateinit var collectionRenderer: CollectionRenderer<UserDetailItem, RecyclerItemAdapter.ViewHolder>

    init {
        SoundCloudApplication.getObjectGraph().inject(this)
    }

    override fun getScreen(): Screen = Screen.USER_INFO

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        collectionRenderer = CollectionRenderer(adapterFactory.create(followersClickListener, followingsClickListener, linkClickListener),
                                                sameIdentity = { firstItem, secondItem -> firstItem.javaClass == secondItem.javaClass },
                                                emptyStateProvider = null)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        collectionRenderer.attach(view, false, LinearLayoutManager(view.context))
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onDestroyView() {
        collectionRenderer.detach()
        super.onDestroyView()
    }

    override fun requestContent(): Observable<UserDetailsParams> {
        return Observable.just(userDetailsParams())
    }

    private fun userDetailsParams(): UserDetailsParams {
        return UserDetailsParams(
                userUrn = requireNotNull(Urns.urnFromBundle(arguments, ProfileArguments.USER_URN_KEY)) { "Missing required param ${ProfileArguments.USER_URN_KEY}" },
                searchQuerySourceInfo = arguments.getParcelable(ProfileArguments.SEARCH_QUERY_SOURCE_INFO_KEY)
        )
    }

    override fun accept(loaderState: AsyncLoaderState<List<UserDetailItem>>) = with(loaderState) {
        val alwaysShowFollows = (data ?: UserDetailItem.Companion.empty(userDetailsParams())) + listOf(UserLoadingItem).filter { asyncLoadingState.isLoadingNextPage }
        collectionRenderer.render(CollectionRendererState(AsyncLoadingState(), alwaysShowFollows))
    }

    override fun disconnectPresenter(presenter: UserDetailsPresenter) = presenter.detachView()

    override fun connectPresenter(presenter: UserDetailsPresenter) = presenter.attachView(this)

    override fun createPresenter() = presenterLazy.get()

    override fun refreshSignal() = collectionRenderer.onRefresh().map { userDetailsParams() }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.recyclerview_with_refresh_without_empty, container, false)

    companion object {
        fun create(userUrn: Urn, searchQuerySourceInfo: SearchQuerySourceInfo?): UserDetailsFragment {
            return UserDetailsFragment().apply {
                arguments = Bundle().apply {
                    Urns.writeToBundle(this, ProfileArguments.USER_URN_KEY, userUrn)
                    putParcelable(ProfileArguments.SEARCH_QUERY_SOURCE_INFO_KEY, searchQuerySourceInfo)
                }
            }
        }
    }
}
